### PR TITLE
Sharder rework

### PR DESCRIPTION
**Pull request by neal-erickson**
Using a new approach of tracking "last shard" for various classical and quantum hazards rather than scanning all preexisting shards when determining dependencies.

### CODE REVIEW

File: pytket/phir/sharding/sharder.py [hunk 1]
```diff
@@ -29,13 +29,6 @@
 logger = logging.getLogger(__name__)
 
 
-def _is_command_global_phase(command: Command) -> bool:
```
Comment (neal-erickson) [inline pytket/phir/sharding/sharder.py:32]: Moved to be a static method for cleanliness.

File: pytket/phir/sharding/sharder.py [hunk 2]
```diff
@@ -135,32 +133,9 @@ def _build_shard(self, command: Command) -> None:
...
-            # Check for read-after-write (value seen would change if reordered)
-            elif not shard.bits_written.isdisjoint(bits_read):
-                logger.debug("...adding shard dep %s -> RAW", shard.ID)
-                depends_upon.add(shard.ID)
-
-            # Check for write-after-read (no reordering or read is changed)
-            elif not shard.bits_written.isdisjoint(bits_read):
-                logger.debug("...adding shard dep %s -> WAR", shard.ID)
-                depends_upon.add(shard.ID)
+        depends_upon = self._resolve_shard_dependencies(
```
Comment (neal-erickson) [inline pytket/phir/sharding/sharder.py:136]: Moved logic into method

File: pytket/phir/sharding/sharder.py [hunk 3]
```diff
@@ -170,16 +145,93 @@ def _build_shard(self, command: Command) -> None:
...
+        logger.debug(
+            "Resolving shard dependencies with qubits=%s bits_written=%s bits_read=%s",
+            qubits,
+            bits_written,
+            bits_read,
+        )
+
+        depends_upon: set[int] = set()
+
+        for qubit in qubits:
```
Comment (neal-erickson) [inline pytket/phir/sharding/sharder.py:177]: These iterations should be extremely fast as they only have to iterate the bits/qubits of the command in question. Could be optimized slightly further but I am almost certain not worth the effort versus the explicitness here.

**Comment by qartik** [comment]
@neal-erickson do you know if this PR includes a fix for #61?

**Review by qartik — APPROVED** [review]
LGTM

**Comment by neal-erickson** [comment]
@qartik No, will address next

**Pull request merged**